### PR TITLE
feat: add operation mapping for device profile

### PR DESCRIPTION
**Pull request by Ruadhri17**
## Proposed changes
TODO:

- [x] test for main device mapping
- [x] test for child device mapping
- [x] doc test for c8y deserializer
- [x] test deserilaizer 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
- #2921 
- #2920 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/c8y_api/src/json_c8y_deserializer.rs [hunk 1]
```diff
@@ -413,6 +424,77 @@ pub struct C8yFirmware {
...
+    }
+
+    pub fn get_config_info(&self) -> Vec<ConfigInfo> {
+        let mut config_info = Vec::new();
+        for config in &self.configuration {
+            config_info.push(ConfigInfo {
+                config_type: config.config_type.clone(),
+                remote_url: Some(config.url.clone()),
+            })
+        }
```
Comment (rina23q) [inline crates/core/c8y_api/src/json_c8y_deserializer.rs:492]: This part can be implemented as `From<C8yDownloadConfigFile> for ConfigInfo`.

**Review by rina23q — COMMENTED** [review]
Some suggestions

File: crates/extensions/c8y_mapper_ext/src/converter.rs [hunk 2]
```diff
@@ -686,6 +686,24 @@ impl CumulocityConverter {
                     vec![]
                 }
             }
+            C8yDeviceControlOperation::DeviceProfile(request) => {
+                if self.config.capabilities.device_profile {
+                    if let Some(profile_name) = extras.get("profileName") {
```
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:691]: Though, `profileName` is not a part of `C8yDeviceProfile` but `C8yOperation`  and it looks like this:
```
"id": "123456",
"profileName": "test-profile",
"c8y_DeviceProfile": {
    /*...*/
},
"externalSource": {
    /*...*/
}
```
I could merge those two json values into one and pass to `C8yDeviceProfile` struct but not sure it is viable solution. It would look like this:
```rust
else if let (Some(c8y_value), Some(name_value)) =
            (hashmap.get("c8y_DeviceProfile"), hashmap.get("profileName"))
        {
            let c8y_value_with_name = {
                c8y_value
                    .clone()
                    .as_object_mut()
                    .unwrap()
                    .append(name_value.clone().as_object_mut().unwrap());
                ().into()
            };
            C8yDeviceControlOperation::DeviceProfile(C8yDeviceProfile::from_json_value(
                c8y_value_with_name,
            )?)
```

File: crates/core/c8y_api/src/json_c8y_deserializer.rs [hunk 3]
```diff
@@ -413,6 +424,77 @@ pub struct C8yFirmware {
     pub url: String,
 }
```
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply crates/core/c8y_api/src/json_c8y_deserializer.rs:480]: That was supposed to be done before the weekend but didn't manage to push the changes. Solved by 5efc5fb95343b4624fab8176b25e6a48dd9da66f

File: crates/core/c8y_api/src/json_c8y_deserializer.rs [hunk 4]
```diff
@@ -413,6 +424,77 @@ pub struct C8yFirmware {
...
+}
+
+impl C8yDeviceProfile {
+    pub fn get_firmware_info(&self) -> FirmwareInfo {
+        FirmwareInfo {
+            name: Some(self.firmware.name.clone()),
+            version: Some(self.firmware.version.clone()),
+            remote_url: Some(self.firmware.url.clone()),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply crates/core/c8y_api/src/json_c8y_deserializer.rs]: Solved by 5efc5fb95343b4624fab8176b25e6a48dd9da66f

File: crates/core/c8y_api/src/json_c8y_deserializer.rs [hunk 5]
```diff
@@ -413,6 +424,77 @@ pub struct C8yFirmware {
...
+
+    pub fn try_get_software_info(&self) -> Result<SoftwareInfo, C8yJsonOverMqttDeserializerError> {
```
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply crates/core/c8y_api/src/json_c8y_deserializer.rs]: Solved by 5efc5fb95343b4624fab8176b25e6a48dd9da66f

File: crates/core/c8y_api/src/json_c8y_deserializer.rs (hunk 1 above)
Comment (Ruadhri17) [reply crates/core/c8y_api/src/json_c8y_deserializer.rs]: Resolved by 5efc5fb95343b4624fab8176b25e6a48dd9da66f

File: crates/core/tedge_api/src/device_profile_command.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
+use crate::commands::Command;
```
Comment (albinsuresh) [inline crates/core/tedge_api/src/device_profile_command.rs:1]: I'd drop the `_command` suffix and just name this module `device_profile` so that it can be moved under a `commands` module directory later, where other individual command structs from the existing `commands.rs` module can also be extracted into individual modules. Then, we wouldn't want each operation module to end with `_command` suffix.

File: crates/extensions/c8y_mapper_ext/src/operations/device_profile.rs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        // Simulate c8y_DeviceProfile operation delivered via JSON over MQTT
+        mqtt.send(MqttMessage::new(
+            &C8yDeviceControlTopic::topic(&"c8y".try_into().unwrap()),
+            json!({
+                "id": "123456",
+                "profileName": "test-profile",
+                "c8y_DeviceProfile": {
+                    "software": [
+                    {
+                        "softwareType": "apt",
```
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/operations/device_profile.rs:85]: Drop the `softwareType` from tests as it is not there at the moment. Use `::apt` type suffixes in the version value for any `type` related tests.

**Review by albinsuresh — CHANGES_REQUESTED** [review]
The mapping logic looks fine in general. But spotted a few corner cases.

The device profile payload may or may not contain all of `software`, `firmware` and `configuration` values at all times. If `firmware` is not included, it won't even be there in the cloud payload. But the current logic expects that key at all times, failing to convert otherwise. `software` and `configuration` can be empty arrays as well but not absent. But, I'd be defensive and code against their absence as well. Also, it'd also be nice if those entries are excluded from the mapped tedge command payload when those arrays are empty as well, just to reduce noise. But that'd be a secondary improvement.

Also, improve the test coverage for `softwareType` conversions as well, by using type suffixes in versions.

File: crates/core/c8y_api/src/json_c8y_deserializer.rs [hunk 8]
```diff
@@ -405,14 +470,72 @@ pub struct C8yDownloadConfigFile {
...
+///     },
+///     "software": [
+///         {
+///             "softwareType": "dummy",
+///             "name": "foo",
+///             "action": "install",
+///             "id": "123456",
+///             "version": "2.0.0",
+///             "url": "https://example.cumulocity.com/inventory/binaries/757538"
+///         },
```
[the remark this answers is not in the document]
Comment (reubenmiller) [reply crates/core/c8y_api/src/json_c8y_deserializer.rs:512]: We can still leave in the `softwareType` property though, as long as we don't fail if it is not provided.

File: crates/extensions/c8y_mapper_ext/src/operations/device_profile.rs (hunk 7 above)
Comment (reubenmiller) [reply crates/extensions/c8y_mapper_ext/src/operations/device_profile.rs:85]: We can write tests for both (with and without softwareType)...as there is nothing preventing people from creating the device profiles via REST API (instead of the UI).

File: crates/core/c8y_api/src/json_c8y_deserializer.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply crates/core/c8y_api/src/json_c8y_deserializer.rs:512]: That `id` is a mistake from dummy copying and should not be there. I would also keep the `softwareType`.

File: crates/core/tedge_api/src/device_profile_command.rs (hunk 6 above)
Comment (Ruadhri17) [reply crates/core/tedge_api/src/device_profile_command.rs]: Resolved by b95a47342e537c2fe0e65d8fb382615ed36ab1e8

File: crates/core/c8y_api/src/json_c8y_deserializer.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply crates/core/c8y_api/src/json_c8y_deserializer.rs]: Resolved by 4db0a80e835a67628e9db7542656ccec18fffa23

File: crates/extensions/c8y_mapper_ext/src/operations/device_profile.rs (hunk 7 above)
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/operations/device_profile.rs:86]: Resolved by 98b7ffbccc7b8118f7a45d4ef553f28b823bfaba

**Review by albinsuresh — COMMENTED** [review]
The only major missing piece is the status update mapping.

Although the test coverage has improved, a few more cases can be added as highlighted. Also, note that the codecov report isn't favourable either. You'll have to make it happy as well.

File: crates/extensions/c8y_mapper_ext/src/converter.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:691]: > Though, profileName is not a part of C8yDeviceProfile but C8yOperation and it looks like this:

I don't get your comment. I thought `C8yDeviceProfile` is a representation of `c8y_DeviceProfile` operation. What I am saying is like this:

```rust
#[derive(Debug, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct C8yDeviceProfile {
    pub profile_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub firmware: Option<C8yFirmware>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub software: Option<C8ySoftwareUpdate>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub configuration: Vec<C8yDownloadConfigFile>,
}
```

Does serde deserilization doesn't work with this?

[the remark this answers is not in the document]
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:691]: Exactly, I was also disappointed with current solution.

File: crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+                    "name": "test-firmware",
+                    "version": "1.0",
+                    "url": "http://www.my.url"
+                    }
+                },
+                "externalSource": {
+                    "externalId": "test-device",
+                    "type": "c8y_Serial"
+                }
+            })
```
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs]: Resolved by b2ca6c501131318afe95ac90b8b9c1249908170b

File: crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+                            "remoteUrl":"http://www.my.url"
+                        }
+                        }
+                    ]
+                }),
+            )],
+        )
+        .await;
+    }
+}
```
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs:785]: Resolved by b2ca6c501131318afe95ac90b8b9c1249908170b

File: crates/extensions/c8y_mapper_ext/src/operations/handlers/mod.rs [hunk 11]
```diff
@@ -134,6 +135,9 @@ impl OperationContext {
                 self.handle_firmware_update_state_change(&entity, &cmd_id, &message)
                     .await
             }
+            OperationType::DeviceProfile => {
+                Ok(OperationOutcome::Ignored) // to do handle state change for device profile
```
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply crates/extensions/c8y_mapper_ext/src/operations/handlers/mod.rs]: Resolved by ea4a5307fb9d4f53c084cf9e29161918cecfba7

File: crates/extensions/c8y_mapper_ext/src/operations/handlers/mod.rs [hunk 12]
```diff
@@ -258,6 +263,7 @@ fn to_c8y_operation(operation_type: &OperationType) -> Option<CumulocitySupporte
         OperationType::SoftwareList => None,
         // local-only operation, not always invoked by c8y, handled in other codepath
         OperationType::Health => None,
+        OperationType::DeviceProfile => Some(CumulocitySupportedOperations::C8yDeviceProfile),
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/operations/handlers/mod.rs:266]: Minor: Moving this line between `OperationType::SoftwareUpdate` and `OperationType::SoftwareList` will look better. Then the operation types at upper part return `Some(_)`, and others at bottom return `None`.

File: crates/extensions/c8y_mapper_ext/src/tests.rs [hunk 13]
```diff
@@ -2786,6 +2786,7 @@ pub(crate) fn test_mapper_config(tmp_dir: &TempTedgeDir) -> C8yMapperConfig {
         config_snapshot: true,
         config_update: true,
         firmware_update: true,
+        device_profile: true,
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/tests.rs:2789]: If this one was using `Capabilities::default()`, you wouldn't need to add a new line...


```suggestion
    let capabilities = Capabilities::default();
```

File: crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs [hunk 14]
```diff
@@ -0,0 +1,1208 @@
...
+            None => {
+                // The command has been fully processed
+                return Ok(OperationOutcome::Ignored);
+            }
+        };
+
+        let sm_topic = &target.smartrest_publish_topic;
+
+        match command.status() {
+            CommandStatus::Executing => Ok(OperationOutcome::Executing),
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs:43]: Also for `Executing`, you need to send a SmartREST message `121,false`. Refer to the [doc](https://cumulocity.com/docs/device-integration/fragment-library/#device-profile)

> ...
> 2. Set operation status to EXECUTING
> 501,c8y_DeviceProfile
> 3. Set target profile
> 121,false,
> ...

Comment (rina23q) [reply crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs:43]: Also, since it's a SmartREST message that updates `inventory`, I recommend you to add a function to the `crates/core/c8y_api/src/smartrest/inventory.rs` file and call the function for `Executing` and `Succssful`.

File: crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs [hunk 15]
```diff
@@ -0,0 +1,1208 @@
...
+                    }
+                ]
+            })
+            .to_string(),
+        ))
+        .await
+        .expect("Send failed");
+
+        // Expect `503` smartrest message on `c8y/s/us`.
+        assert_received_contains_str(&mut mqtt, [("c8y/s/us/child1", "503,c8y_DeviceProfile")])
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs:1205]: Check for `121,true` message?

File: crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs [hunk 16]
```diff
@@ -0,0 +1,1208 @@
...
+                    }
+                ]
+            })
+            .to_string(),
+        ))
+        .await
+        .expect("Send failed");
+
+        // Expect `501` smartrest message on `c8y/s/us`.
+        assert_received_contains_str(&mut mqtt, [("c8y/s/us", "501,c8y_DeviceProfile")]).await;
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs:854]: Need a check for `121,false` message after you implement it.

File: crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs [hunk 17]
```diff
@@ -0,0 +1,1208 @@
...
+                    }
+                ]
+            })
+            .to_string(),
+        ))
+        .await
+        .expect("Send failed");
+
+        // Expect `501` smartrest message on `c8y/s/us`.
+        assert_received_contains_str(&mut mqtt, [("c8y/s/us/child1", "501,c8y_DeviceProfile")])
```
Comment (rina23q) [inline crates/extensions/c8y_mapper_ext/src/operations/handlers/device_profile.rs:993]: Also here needs `121,false` check

**Review by rina23q — COMMENTED** [review]
Overall looks good, but still some points need to be changed. Especially `121,false` message is missing for the `Executing` state.

**Comment by rina23q** [comment]
While I am testing end-to-end, I found that handling for the `device_profile` command metadata is missing. This means, `c8y_DeviceProfile` is never declared in `114` message...

This match needs to be expanded. The implementation should be very similar to the case `OperationType::Restart`.
https://github.com/thin-edge/thin-edge.io/blob/1c68fab17eafe75dbefefbe6b7caf02ff313348a/crates/extensions/c8y_mapper_ext/src/converter.rs#L1123-L1149

I'll let you decide if you address this in this PR or a follow-up PR.

**Comment by rina23q** [comment]
Another problem: c8y-mapper doesn't subscribe `../cmd/device_profile` topic and `../cmd/device_profile/+` topic. You have to add something like below

```
 if capabilities.device_profile { 
     topics.add_all(crate::operations::device_profile::topic_filter(&mqtt_schema)); 
 } 
```

to this file

https://github.com/thin-edge/thin-edge.io/blob/1c68fab17eafe75dbefefbe6b7caf02ff313348a/crates/extensions/c8y_mapper_ext/src/config.rs#L201-L218

**Review by rina23q — APPROVED** [review]
I confirmed both issue were fixed. Thank you.

**Review by albinsuresh — APPROVED** [review]
The feature is working as expected. But, the codecov report is not at all favourable. Something needs to be done to get it in shape.

**Comment by gligorisaev** [comment]
LGTM

**Comment by gligorisaev** [comment]
The feature is working as expected, QA checked the functionality according description
